turbopack/crates/turbo-rcstr/src/dynamic.rs

```diff
@@ -0,0 +1,76 @@
+use std::{
+    borrow::Cow,
+    hash::{Hash, Hasher},
+    ptr::NonNull,
+};
+
+use triomphe::Arc;
+
+use crate::{
+    tagged_value::{TaggedValue, MAX_INLINE_LEN},
+    RcStr, INLINE_TAG_INIT, LEN_OFFSET, TAG_MASK,
+};
+
+#[derive(Debug)]
+pub(crate) struct Entry {
+    pub string: Arc<str>,
+}
+
+impl Entry {
+    pub unsafe fn cast(ptr: TaggedValue) -> *const Entry {
+        ptr.get_ptr().cast()
+    }
+
+    pub unsafe fn deref_from<'i>(ptr: TaggedValue) -> &'i Entry {
+        &*Self::cast(ptr)
+    }
+
+    pub unsafe fn restore_arc(v: TaggedValue) -> Arc<Entry> {
+        let ptr = v.get_ptr() as *const Entry;
+        Arc::from_raw(ptr)
+    }
+}
+
+impl PartialEq for Entry {
+    fn eq(&self, other: &Self) -> bool {
+        self.string == other.string
+    }
+}
+
+impl Eq for Entry {}
+
+impl Hash for Entry {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.string.hash(state);
+    }
+}
+
+/// This can create any kind of [Atom], although this lives in the `dynamic`
+/// module.
+pub(crate) fn new_atom(text: Cow<str>) -> RcStr {
```

Comment:
As this is likely really hot code, this might be better as `text: impl Into<String>` because the code could be specialized for `String` vs `&str` vs `Cow<str>`. Just using `Cow<str>` here may require a runtime branch.

Though... this might get inlined anyways, and so generics might not be needed to perform specialization.